### ISSUE TITLE
Compilation with gcc5 fails due to ABI break

### ISSUE THREAD

**Issue by vosst** [mapbox/mapbox-gl-native#3123]
When trying to build mapbox-gl-native with gcc/g++5 or later, the build fails with a linker error:
```
./Release/obj.target/gyp/libmbgl-core.a(shape_annotation_impl.o): In function `mbgl::ShapeAnnotationImpl::updateTile(mbgl::TileID const&, mbgl::AnnotationTile&)':
/tmp/mapbox-gl-native/build/linux-x86_64/../../src/mbgl/annotation/shape_annotation_impl.cpp:108: undefined reference to `mapbox::util::geojsonvt::Convert::create(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, mapbox::util::geojsonvt::ProjectedFeatureType, mapbox::util::variant<mapbox::util::geojsonvt::ProjectedPoint, mapbox::util::geojsonvt::ProjectedGeometryContainer>)'
```
The issue is likely caused by the ABI break introduced by gcc/g++5 for supporting C++11 and the respective adjustments to std::string/std::list. The respective symbol names are mangled with a std::__cxx11 prefix. The geojsonvt static library managed via mason is likely built with gcc/g++ 4.9 and the required symbols are not named correctly and thus not resolved by the linker.

One possible solution could be to force gcc 4.9 in the build system setup and error out if gcc 4.9 is not available on the build machine.

**Comment by mikemorris** [comment]
The mason `GeoJSONVT` lib is actually built with clang, so yes, an ABI conflict is possible. We were able to work around this in GCC 4.9 builds by setting [`-fabi-version=0`](https://github.com/mapbox/mapbox-gl-native/blob/146533e377ad5ae205a52d37f98771da3f39a547/gyp/common.gypi#L50) ("newest"), which [should be the default on GCC5](https://github.com/mapbox/mapbox-gl-native/issues/2502#issuecomment-146753591). Can you confirm that this flag is being set in your build, and the exact version of GCC you're using?

https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html

We're tracking this issue as it relates to mason packages over in https://github.com/mapbox/mason/issues/111

**Comment by jfirebaugh** [comment]
https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

Although mason GeoJSONVT was compiled with clang against libstdc++, it's probably not complied with `_GLIBCXX_USE_CXX11_ABI` set. We may need to change that.

**Comment by vosst** [comment]
I'm using gcc (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010 (@Ubuntu wily).
The flag is set (output from `V=1 make linux`):
```
g++ '-DNDEBUG' -I../../include -I../../src -I./Release/obj/gen/include -I./Release/obj/gen/include  -I/tmp/mapbox-gl-native/mason_packages/linux-x86_64/libuv/1.7.5/include -I/usr/include/libdrm -isystem /tmp/mapbox-gl-native/mason_packages/headers/variant/1.0/include -I/tmp/mapbox-gl-native/mason_packages/headers/variant/1.0/include -I/tmp/mapbox-gl-native/mason_packages/headers/rapidjson/1.0.2/include -fPIC -std=c++14 -Werror -Wall -Wextra -Wshadow -Wno-variadic-macros -Wno-error=unused-parameter -frtti -fexceptions  -Wno-unknown-pragmas -fabi-version=0 -I/tmp/mapbox-gl-native/mason_packages/linux-x86_64/libuv/1.7.5/include -I/usr/include/libdrm -I/tmp/mapbox-gl-native/mason_packages/headers/boost/1.59.0/include -I/tmp/mapbox-gl-native/mason_packages/linux-x86_64/geojsonvt/2.1.6.3/include -isystem /tmp/mapbox-gl-native/mason_packages/headers/variant/1.0/include -I/tmp/mapbox-gl-native/mason_packages/headers/variant/1.0/include -I/tmp/mapbox-gl-native/mason_packages/headers/rapidjson/1.0.2/include -fPIC -g -O3 -MMD -MF ./Release/.deps/Release/obj.target/core/src/mbgl/shader/circle_shader.o.d.raw  -c -o Release/obj.target/core/src/mbgl/shader/circle_shader.o ../../src/mbgl/shader/circle_shader.cpp
```

I'm not sure whether clang supports dual abis at all (see http://clang-developers.42468.n3.nabble.com/GCC-5-1-ABI-GLIBCXX-USE-CXX11-ABI-td4045387.html). For Ubuntu, we pretty much recompiled the C++ world when handling the GCC 5 transition.

**Comment by nuket** [comment]
Seeing this issue as well on an Ubuntu 16.04 xenial docker image, set up using the following commands from the Linux setup README and the Qt setup README.

Instead of libllvm3.4, I had to use the `llvm` package, since llvm 3.4 isn't available on 16.04.

```
sudo apt-get install software-properties-common
sudo add-apt-repository --yes ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt-get install gcc-5 g++-5
export CXX=g++-5

sudo apt-get install curl git build-essential zlib1g-dev automake \
                     libtool xutils-dev make cmake pkg-config python-pip \
                     libcurl4-openssl-dev libpng-dev libsqlite3-dev \
                     libllvm3.4

sudo apt-get install libxi-dev libglu1-mesa-dev x11proto-randr-dev \
                     x11proto-xext-dev libxrandr-dev \
                     x11proto-xf86vidmode-dev libxxf86vm-dev \
                     libxcursor-dev libxinerama-dev

sudo apt-get install qt5-default qtlocation5-dev qtpositioning5-dev \
    qml-module-qtquick2 qml-module-qtpositioning qml-module-qtgraphicaleffects \
    qml-module-qtquick-layouts qml-module-qtquick-controls

CXX=g++-5 make qt-qml-app
```

```
max@28f58ff6ee4d:~/devsrc/mapbox-gl-native$ CXX=g++-5 make glfw-app
make -f scripts/main.mk Makefile/glfw-app
make[1]: Entering directory '/home/max/devsrc/mapbox-gl-native'
* Recreating project...
deps/run_gyp platform/linux/platform.gyp -I./build/linux-x86_64/config.gypi -Dcoverage=0 -Dcxx_host="g++" -Goutput_dir=. --depth=. --generator-output=./build/linux-x86_64 -f make
* Building target glfw-app...
make -j4 -C ./build/linux-x86_64 BUILDTYPE=Release glfw-app
make[2]: Entering directory '/home/max/devsrc/mapbox-gl-native/build/linux-x86_64'
  CXX(target) Release/obj.target/core/src/csscolorparser/csscolorparser.o
  AR(target) Release/obj.target/platform/linux/libmbgl-core.a
  COPY Release/libmbgl-core.a
  AR(target) Release/obj.target/platform/linux/libmbgl-platform-linux.a
  COPY Release/libmbgl-platform-linux.a
  LINK(target) Release/mapbox-glfw
./Release/obj.target/platform/linux/libmbgl-core.a(shape_annotation_impl.o): In function `mbgl::ShapeAnnotationImpl::updateTile(mbgl::TileID const&, mbgl::AnnotationTile&)':
/home/max/devsrc/mapbox-gl-native/build/linux-x86_64/../../src/mbgl/annotation/shape_annotation_impl.cpp:107: undefined reference to `mapbox::geojsonvt::Convert::create(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, mapbox::geojsonvt::ProjectedFeatureType, mapbox::util::variant<std::vector<mapbox::geojsonvt::ProjectedPoint, std::allocator<mapbox::geojsonvt::ProjectedPoint> >, std::vector<mapbox::geojsonvt::ProjectedRing, std::allocator<mapbox::geojsonvt::ProjectedRing> > >)'
collect2: error: ld returned 1 exit status
platform/linux/glfw-app.target.mk:180: recipe for target 'Release/mapbox-glfw' failed
make[2]: *** [Release/mapbox-glfw] Error 1
make[2]: Leaving directory '/home/max/devsrc/mapbox-gl-native/build/linux-x86_64'
scripts/main.mk:123: recipe for target 'Makefile/glfw-app' failed
make[1]: *** [Makefile/glfw-app] Error 2
make[1]: Leaving directory '/home/max/devsrc/mapbox-gl-native'
Makefile:219: recipe for target 'glfw-app' failed
make: *** [glfw-app] Error 2
max@28f58ff6ee4d:~/devsrc/mapbox-gl-native$ 
```

```
max@28f58ff6ee4d:~/devsrc/mapbox-gl-native$ g++-5 -v
Using built-in specs.
COLLECT_GCC=g++-5
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.3.1-14ubuntu2' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2) 
```

```
max@28f58ff6ee4d:~/devsrc/mapbox-gl-native$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04 LTS
Release:	16.04
Codename:	xenial
```

**Comment by tmpsantos** [comment]
@nuket try building the `tmpsantos-cxx11abi` branch, it should work.

I'm working on a proper solution for this to push to the master branch, but so far it includes using `readelf` and looking up for `::cxx11` symbols on the `libstdc++` and I'm not very happy with it.
